accounting/invoice/models.py

```diff
@@ -20,3 +20,387 @@
+        )
+        pdf_path = os.path.join(settings.INVOICE_PDF_PATH, pdf_name)
+        pdf_configuration = pdfkit.configuration(wkhtmltopdf=settings.HTML_TO_PDF_BINARY_PATH)
+        pdfkit.from_string(invoice, pdf_path, configuration=pdf_configuration, options=self.PDF_OPTIONS)
+        return pdf_path
+
+    def upload_to_google_drive(self, invoice_file):  # NOQA
+        """
+        Upload the invoice in some file format to Google Drive.
+
+        Assumes the invoice has already been turned into a file, i.e. a PDF.
+        Assumes a _very_ specific file structure right now. For example:
+            |- 2016
+            |- 2017
+            |   |- blah
+            |   |- ...
+            |   |- invoices-in
+            |   |   |- 1
+            |   |   |- 2
+            |   |   |- ...
+            |   |   |- 12
+            |   |   |   |- invoice_1.pdf
+            |   |   |   |- ...
+            |   |   |   |- invoice_n.pdf
+            |- 2018
+
+        TODO: This should be a synchronous job on a non-web worker.
+        """
+        if not settings.ENABLE_GOOGLE:
+            return
+
+        # Generate a query we'll use repeatedly for finding the proper folder.
+        query = ("mimeType ='application/vnd.google-apps.folder' and "
+                 "'{parent}' in parents and "
+                 "trashed=false")
+
+        def get_folder_id(root, path=None):
+            """
+            Return the ID of the folder with the last name in `path`, starting traversal from folder given by ID `root`.
+            """
+            if path and isinstance(path, list):
+                folders = drive.ListFile({'q': query.format(parent=root)}).GetList()
+                for folder in folders:
+                    if folder['title'] == path[0]:
+                        root = folder['id']
+                        break
+                return get_folder_id(root, path[1:])
+            return root
+
+        gauth = GoogleAuth()
+        gauth.ServiceAuth()
+        drive = GoogleDrive(gauth)
+        month_folder = get_folder_id(
+            settings.GOOGLE_DRIVE_ROOT,
+            path=[str(self.billing_start_date.year), 'invoices-in', str(self.billing_start_date.month)]
+        )
+        file = drive.CreateFile({
+            'title': invoice_file.split('/')[-1].split('|')[-1],  # TODO: Make this cleaner.
+            'parents': [{'id': month_folder}]
+        })
+        file.SetContentFile(invoice_file)
+        file.Upload()
+
+    def save(self, **kwargs):
+        """
+        Convert the invoice to a PDF and/or upload it to Google Drive after a successful save.
+        """
+        super().save(**kwargs)
+        if self.auto_download_jira_worklogs_on_save:
+            self.fill_line_items_from_jira()
+        if self.auto_create_pdf_on_save:
+            pdf = self.to_pdf()
+            if self.auto_upload_google_drive_on_save:
+                self.upload_to_google_drive(pdf)
+
+
+class LineItem(models.Model):
```

Comment:
Worth splitting out Invoice and LineItem to separate files?

Reply:
It's going to cause cyclic imports.